Make access to parameters in sections also require full path

bernt-matthias opened the pull request:
So far the access to values of parameter in sections did not require to specify the full access path.

For

```
<section name="sec">
   <param name="par">
</section>
```

in format_source just `par` was used. With the change the access is format_source `sec|par`.

In cheetah `$sec.par` is already used.

The problem was discovered for an automatic tool conversion where two parameters with the same name existed where one was in a section and the other wasn't. Then one did overwrite the other.

bernt-matthias committed:
format_source: changed test to show bug

format_source seems dysfunctional for inputs with the same name
exist in different sections

mvdbeek commented:
That's a nice test, fixing this probably also fixes https://github.com/galaxyproject/galaxy/issues/9486

bernt-matthias commented:
Cool. Do you know if this worked in the past and just broke or if this never worked?

mvdbeek commented:
You're not supposed to duplicate parameter names (if they're not the same thing, as in conditional switches), so there is a chance it never worked, but some years ago we also reworked the param dict lookup for unqualified parameter names that could have affected this.

bernt-matthias commented:
I understand and would never do this when crating wrappers manually.

But, I stumbled over this when autogenerating wrappers for OpenMS .. so I don't have any influence on the parameter names .. or need quite a bit of additional code which I would like to avoid.

mvdbeek commented:
sure, and that's why I like your test. clearly we don't want to mask `input_data_1 ` with `test_section .input_data_1 `

bernt-matthias commented:
Good thing is that the test fails: `Dataset metadata verification for [file_ext] failed, expected [fastqsanger] but found [fastqsolexa]`

But from my test that brought be to this bug I would have expected `Dataset metadata verification for [file_ext] failed, expected [fastqsanger] but found [data]`

bernt-matthias committed:
fix handling of sections in visit_input_values

the parent_prefix was apparently set wrong

API tests have been adapted to include one case of a Section
(one repeat has been replaced)

bernt-matthias commented:
So here is a possible fix .. at least this test is working now as expected. Maybe this will have side effects, since inputs in sections are now referenced as `section|input`.

Not sure if this is actually a bug in the visitor:

- somehow the `parent_prefix` is empty when it probably shouldn't so I return `name_prefix` https://github.com/galaxyproject/galaxy/blob/972fd28601dbb8815d51e6fd4118c948ee2ae314/lib/galaxy/tools/parameters/__init__.py#L128
- alternatively the calling functions could use `prefixed_name`

bernt-matthias committed:
fix handling of conditionals in visit_input_values

same as for sections

API tests have been adapted

bernt-matthias committed:
also use qualified name for structured_like in test

bernt-matthias commented:
Hey @mvdbeek .. this gets a bit tricky and I could need a bit of feedback. 

Seemingly `visit_input_values` did not add prefixes for sections and conditionals so far. That's why the input in the new test was masked. My changes would change this and require that the full name with prefixes is necessary for all attributes referring to inputs (so far I'm aware of `format_source` and `structured_like`. 

I guess this would break a few tools in its current state (quick check at IUC confirms this). 

Also there are some side effects which I do not understand yet (e.g. the py3 api tests).

I'm thinking about a good way to fix this. 

- one could add both the qualified and the unqualified name here https://github.com/galaxyproject/galaxy/blob/9ba1f4550cfa9d25074dd0764fa7189aed10eb36/lib/galaxy/tools/actions/__init__.py#L64
 
- I also remember that in tests with sections / conditionals it is possible address the inputs with and without the section or conditional name (i.e. `section|id` or just `id`). But I'm not sure if I'm right. If so then maybe we can add the same flexibility here in the same way?

What's your take on this?

bernt-matthias committed:
try to fix api tests

bernt-matthias committed:
c

bernt-matthias committed:
start hiding new functionality behind profile

bernt-matthias committed:
extend test for visit_input_values

to both profiles

bernt-matthias committed:
present profile as string

bernt-matthias committed:
make unit tests visit_input_values consider both profiles

bernt-matthias committed:
add profile to mock tool used in workflow tests

bernt-matthias committed:
restore original functional tool tests

and add new ones testing the profile version for 21.01
that requires qualified access to parameters in secions
and conditionals

bernt-matthias committed:
update galaxy.xsd

to cover that references to inputs from outputs need to be fully qualified

bernt-matthias committed:
undo changes to identifier_in_conditional.xml

bernt-matthias committed:
fix collection_paired_conditional_structured_like_qualified

bernt-matthias committed:
undo changes to workflow api tests

because test currently covers only old behavior
